src/messages/variant.rs

```diff
@@ -60,6 +60,11 @@ pub enum Variant {
     ParsecRequest(u64, parsec::Request),
     /// Parsec response message
     ParsecResponse(u64, parsec::Response),
+    /// Message sent to a lost peer to determine whether the disconnect is permanent or only
+    /// temporary.
+    PingRequest,
```

Comment:
So I'm not clear why we need the PingRequest/Response?
I mean with previous change we get it lost if we cannot send a message to it trying multiple times (which i though was already counted as the acceptable margin for a node to be unresponsive). Why do we have this additional one?

Reply:
To give the node chance to come back. The unsent message thing might be exhausted too quickly.

Reply:
Why should it be exhausted too quickly, is it not what we should ensure is not exhausted too quickly instead of adding a new mechanism. (i.e ensure messages are passed through to all needed targets).

Reply:
The ping mechanism allow us to control the time interval for which we allow a node to be non-responsive. With the `UnsentUserMessage` mechanism we don't have that control. In some cases it can cycle through all the attempts (currently 3) too quickly and we might accuse the node as being lost even though it might not really be.

What I have some doubts about however is whether we actually need the `UnsentUserMessage` mechanism for this at all. Depending on how quic-p2p works, we might get `ConnectionFailure` in all the cases we need. In particular:
1. If we are connected to a peer and that peer goes offline, we should always get `ConnectionFailure` 
2. If we are not connected and we sent them a message, quic-p2p should establish the connection automatically, but if it fails to do so because the peer went offline it **should** emit `ConnectionFailure` too, but I'm not sure this is the case. I think it wasn't the case in the past, but maybe that changed.

Reply:
- Well if we want a minimum time for retry we should add that i.e the retry is not just a number but a time elapsed.

- Regarding ConnectionFailure, yes, I think we should get this, but we will get this on the first fail attempt, I think, I think we fixed at least the mock to do the right thing and give an error. I think the real code was fixed too.

Reply:
OK. I'm going to verify the quic-p2p behaviour regarding this and if it's what we want then I'll remove the unsent mechanism.

Reply:
Sorry I do not understand how you came to that course of action? why would we remove the unsent mechanism.

Reply:
By remove the unsent mechanism I meant don't use it for detecting lost peers. Not removing it completely.

Reply:
Ok. Still not clear how we would do that (i.e even if we get ConnectionFailure can we use it because it does not say whether it is the first one etc...)

Reply:
So there are two phases in this:
1. Detect that a node is lost, which means at least temporarily offline
2. Detect that it is gone, which means it is permanently offline and should be voted out of the section.

This discussion is mainly about point 1. This PR currently implements two mechanisms for that:
1. When we receive the `ConnectionFailure` event
2. Or when we repeatedly  fail to send a message to that peer 

What I'm trying to argue here is that we might not need the second mechanism because the first one might be sufficient. So we don't care whether it's the first time and so on because we only use it to trigger the second phase which is where we use the ping requests/responses to give the node chance to come back.

Does that make sense?

Reply:
It does to an extent, but it seem to provide something that essentially re-implement in a different way the resend mechanism. i.e the resend is design to send to all its expected target unless a node is gone, then we may only be able to send to part of the targets. (if it is broken we should fix it). It seem the reason to stop trying to re-send to a failing node should be the same as the reason we decide a node should be voted offline.

Reply:
OK, I see what you mean. So maybe we should just modify the resend mechanism and not resend to a previously failed target immediately, but only after some delay (which we increase after every attempt)?

Reply:
That could be, though:
 - I think there is a timeout in quic-p2p when trying to connect to send a message, so there already is a delay (to confirm? / maybe not in mock?).
 - Whatever delay we introduce should not delay sending to other targets if available.
 - Not sure if we want an additional delay or whether we want to keep on retrying until enough time has passed and max try reached.

Reply:
- There is no timeout in mock quic-p2p, but we probably don't need that.
- Retrying until enough time passed sounds like a decent idea and should be simpler to implement. I'll give it a try.